### PR TITLE
[SPARK-24863][SS] Report Kafka offset lag as a custom metrics

### PR DESCRIPTION
**Pull request by arunmahadevan**
## What changes were proposed in this pull request?

This builds on top of SPARK-24748 to report 'offset lag' as a custom metrics for Kafka structured streaming source.

This lag is the difference between the latest offsets in Kafka the time the metrics is reported (just after a micro-batch completes) and the latest offset Spark has processed. It can be 0 (or close to 0) if spark keeps up with the rate at which messages are ingested into Kafka topics in steady state. This measures how far behind the spark source has fallen behind (per partition) and can aid in tuning the application.

## How was this patch tested?

Existing and new unit tests

Please review http://spark.apache.org/contributing.html before opening a pull request.

### CODE REVIEW

**Comment by arunmahadevan** [comment]
@HeartSaVioR @HyukjinKwon @jose-torres @tdas  would you mind taking a look?

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/JsonUtils.scala [hunk 1]
```diff
@@ -95,4 +95,20 @@ private object JsonUtils {
     }
     Serialization.write(result)
   }
+
+  /**
+   * Write per-topic partition lag as json string
+   */
+  def partitionLags(latestOffsets: Map[TopicPartition, Long],
+                    processedOffsets: Map[TopicPartition, Long]): String = {
```
Comment (HyukjinKwon) [inline external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/JsonUtils.scala]: nit:

```scala
def partitionLags(
    latestOffsets: Map[TopicPartition, Long],
    processedOffsets: Map[TopicPartition, Long]): String = {
```

per https://github.com/databricks/scala-style-guide#spacing-and-indentation

Please feel free to address it with other comments later.

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/JsonUtils.scala [hunk 2]
```diff
@@ -95,4 +95,20 @@ private object JsonUtils {
...
+    val result = new HashMap[String, HashMap[Int, Long]]()
```
Comment (HyukjinKwon) [inline external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/JsonUtils.scala]: nit: `HashMap.empty[String, HashMap[Int, Long]]`

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/JsonUtils.scala (hunk 1 above)
Comment (arunmahadevan) [reply external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/JsonUtils.scala]: addressed. would it be possible to add this to scala style checks ?

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/JsonUtils.scala (hunk 2 above)
Comment (arunmahadevan) [reply external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/JsonUtils.scala]: Had followed the style in other parts of the class. Addressed and refactored the other places as well.

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
cc @koeninger as well

**Review by koeninger — APPROVED** [review]

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/JsonUtils.scala (hunk 1 above)
Comment (HyukjinKwon) [reply external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/JsonUtils.scala]: Please go ahead if possible.

**Comment by arunmahadevan** [comment]
@HyukjinKwon , can you take it forward? Appreciate your effort and thanks in advance.

**Comment by HyukjinKwon** [comment]
Let me leave this in few days in case someone has more comments on this.

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**